include/deal.II/grid/manifold_lib.h

```diff
@@ -503,6 +503,60 @@ class TorusManifold : public ChartManifold<dim,3,3>
   double r, R;
 };
 
+
+
+/**
+ * A mapping class that smoothly transitions from different manifold
+ * descriptions on different faces by a so-called transfinite
+ * interpolation. As opposed to setting the manifold description only on the
+ * boundary of a mesh, this approach makes sure that the deformation of the
+ * mesh according to a curved boundary description is applied to more than one
+ * element layer once the mesh gets refined, which is necessary to get
+ * high-order convergence rates (without this class, the error on cells close
+ * to the boundary does not exceed 3 on a circle).
+ *
+ * @note Only instantiated for dim > 1 as a transfinite mapping cannot be
+ * defined in 1D.
+ *
+ * @author Luca Heltai, Martin Kronbichler, 2017
+ */
+template <int dim, int spacedim=dim>
+class TransfiniteInterpolationManifold : public ChartManifold<dim,spacedim,dim>
+{
+public:
+  void initialize (const typename Triangulation<dim,spacedim>::cell_iterator &cell);
+
+  /**
+   * Pull back operation.
+   */
+  virtual Point<dim>
+  pull_back(const Point<spacedim> &p) const;
+
+  /**
+   * Push forward operation.
+   */
+  virtual Point<spacedim>
+  push_forward(const Point<dim> &chart_point) const;
+
+  /**
+   * Gradient.
+   */
+  virtual
+  DerivativeForm<1,dim,spacedim>
+  push_forward_gradient(const Point<dim> &chart_point) const;
+
+private:
+  /**
+   * The bounding points of the manifold description.
+   */
+  Point<spacedim> surrounding_points[GeometryInfo<dim>::vertices_per_cell];
+
+  /**
+   * The manifold descriptions for the surrounding faces.
+   */
+  const Manifold<dim,spacedim> *face_manifolds[GeometryInfo<dim>::faces_per_cell];
```

Comment:
Also, would it be better to use `SmartPointer` to make sure the object pointed to doesn't go away?